File: 410-split-array-largest-sum/split-array-largest-sum.java
class Solution {

    public boolean isValidPartition(int[] nums,int k,int minSum)
    {
        int partitions=1;
        int sum=0;
        for(int i=0;i<nums.length;i++)
        {
            sum+=nums[i];
            if(sum>minSum)
            {
                partitions++;
                sum=nums[i];
            }
            if(partitions>k)
            return false;
        }
        return true;
    }

    public int splitArray(int[] nums,int k) 
    {
        int sum=Arrays.stream(nums).sum();  // Calculate the sum of all elements in nums
        int end=sum;
        int start=Arrays.stream(nums).max().orElse(0);  // Find the maximum element in nums

        int minimumSumSubarray=end;
        while(start<=end)
        {
            int mid=start+(end-start)/2;
            if(isValidPartition(nums,k,mid))
            {
                minimumSumSubarray=mid;
                end=mid-1;
            }
            else
                start=mid+1;
        }  
        return minimumSumSubarray;
    }
}

File: 410-split-array-largest-sum/split-array-largest-sum.cpp
class Solution {
public:

    bool isValid(vector<int>& a,int k,int sum)
    {
        int parts=1;
        int currSum=0;
        for(int i=0;i<a.size();i++)
        {
            currSum+=a[i];
            if(currSum>sum)
            {
                parts++;
                currSum=a[i];
            }

            if(parts>k)
            return false;
        }
        return true;
    }

    int splitArray(vector<int>& a, int k) 
    {
        int sum=0;
        int start=*max_element(a.begin(),a.end());
        int end=accumulate(a.begin(),a.end(),sum);
        int minSum=end;

        while(start<=end)
        {
            int mid=start+(end-start)/2;

            if(isValid(a,k,mid))
            {
                minSum=mid;
                end=mid-1;
            }
            else
                start=mid+1;
        }
        return minSum;
    }
};